src/metric_family.cc

```diff
@@ -0,0 +1,230 @@
+
+  // Save the references to shared memory.
+  custom_metric_family_shm_ = std::move(custom_metric_family_shm);
+  name_shm_ = std::move(name_shm);
+  description_shm_ = std::move(description_shm);
+  shm_handle_ = custom_metric_family_shm.handle_;
+}
+
+std::unique_ptr<MetricFamily>
+MetricFamily::LoadFromSharedMemory(
+    std::unique_ptr<SharedMemoryManager>& shm_pool,
+    bi::managed_external_buffer::handle_t handle)
+{
+  AllocatedSharedMemory<char> custom_metric_family_shm =
+      shm_pool->Load<char>(handle);
+  MetricFamilyShm* custom_metric_family_shm_ptr =
+      reinterpret_cast<MetricFamilyShm*>(custom_metric_family_shm.data_.get());
+  std::unique_ptr<PbString> name_shm = PbString::LoadFromSharedMemory(
+      shm_pool, custom_metric_family_shm_ptr->name_shm_handle);
+  std::unique_ptr<PbString> description_shm = PbString::LoadFromSharedMemory(
+      shm_pool, custom_metric_family_shm_ptr->description_shm_handle);
+
+  return std::unique_ptr<MetricFamily>(
+      new MetricFamily(custom_metric_family_shm, name_shm, description_shm));
+}
+
+MetricFamily::MetricFamily(
+    AllocatedSharedMemory<char>& custom_metric_family_shm,
+    std::unique_ptr<PbString>& name_shm,
+    std::unique_ptr<PbString>& description_shm)
+    : custom_metric_family_shm_(std::move(custom_metric_family_shm)),
+      name_shm_(std::move(name_shm)),
+      description_shm_(std::move(description_shm))
+{
+  custom_metric_family_shm_ptr_ =
+      reinterpret_cast<MetricFamilyShm*>(custom_metric_family_shm_.data_.get());
+  name_ = name_shm_->String();
+  description_ = description_shm_->String();
+  kind_ = custom_metric_family_shm_ptr_->kind;
+  metric_family_request_kind_ =
+      custom_metric_family_shm_ptr_->metric_family_request_kind;
+  metric_family_address_ = custom_metric_family_shm_ptr_->metric_family_address;
+}
+
+void*
+MetricFamily::MetricFamilyAddress()
+{
+  return metric_family_address_;
+}
+
+#ifdef TRITON_PB_STUB
+void
+MetricFamily::AddMetric(std::shared_ptr<Metric> metric)
+{
+  std::lock_guard<std::mutex> lock(metric_map_mu_);
+  metric_map_.insert({metric->MetricAddress(), metric});
+}
+#else
+void*
+MetricFamily::InitializeTritonMetricFamily()
+{
+  TRITONSERVER_MetricKind triton_kind = ToTritonServerMetricKind(kind_);
+  TRITONSERVER_MetricFamily* triton_metric_family = nullptr;
+  THROW_IF_TRITON_ERROR(TRITONSERVER_MetricFamilyNew(
+      &triton_metric_family, triton_kind, name_.c_str(), description_.c_str()));
+  return reinterpret_cast<void*>(triton_metric_family);
+}
+
+TRITONSERVER_MetricKind
+MetricFamily::ToTritonServerMetricKind(const MetricKind& kind)
+{
+  switch (kind) {
+    case COUNTER:
+      return TRITONSERVER_METRIC_KIND_COUNTER;
+    case GAUGE:
+      return TRITONSERVER_METRIC_KIND_GAUGE;
+    default:
+      throw PythonBackendException("Unknown metric kind");
+  }
+}
+
+void
+MetricFamily::ClearTritonMetricFamily()
+{
+  metric_map_.clear();
```

Comment:
does this need to be protected by `metric_map_mu_` ?

Reply:
This line is no longer needed as we already clean the map during the stub stage. We don't pass the `metric_map_` to the Python backend.